src/PortingAssistant.Client.Common/Utils/FileSystemAccess.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace PortingAssistant.Client.Common.Utils
+{
+    public static class FileSystemAccess
+    {
+        private static readonly string[] fileTypesToCheck = { ".csproj", ".cs" };
+
+        /// <summary>
+        /// Checks directory and all content for write access
+        /// </summary>
+        /// <param name="path">Directory path</param>
+        /// <returns>List of items without write access</returns>
+        public static List<string> CheckWriteAccessForDirectory(string path)
+        {
+            var result = (Directory.GetFiles(path)
+                .Where(file => fileTypesToCheck.Contains(Path.GetExtension(file)))
+                .Where(file => !CanWriteFile(file))).ToList();
+            foreach (string subDirectory in Directory.GetDirectories(path))
+            {
+                if (!CanWriteToDirectory(subDirectory))
+                {
+                    result.Add(subDirectory);
+                }
+                result.AddRange(CheckWriteAccessForDirectory(subDirectory));
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Checks csproj and at least one .cs file is writeable
+        /// </summary>
+        /// <param name="path">Project file path</param>
+        /// <returns>True if csproj and at least one .cs file is writeable</returns>
+        public static bool CheckWriteAccessForProject(string path)
+        {
+            return CanWriteFile(path) &&
+                 DirectoryHasWriteableCSharpFile(Path.GetDirectoryName(path));
+        }
+
+        private static bool DirectoryHasWriteableCSharpFile(string directoryPath)
+        {
+            bool fileFound = Directory.GetFiles(directoryPath)
+                .Where(file => Path.GetExtension(file) == ".cs")
+                .Any(file => !CanWriteFile(file));
+
+            return fileFound || 
+                Directory.GetDirectories(directoryPath).Any(subDirectory => DirectoryHasWriteableCSharpFile(subDirectory));
+        }
+
+        private static bool CanWriteFile(string path)
+        {
+            try
+            {
+                using var fs = File.Open(path, FileMode.Open);
+                return true;
```

Comment:
Would the filemode.open be enough for writing to a file? do we need to check a different access?

Reply:
When testing it still hit the unauthorized error with just this, I'm thinking because we didn't specify FileAccess.Read. But I think you're right that we should be more specific in trying to open the file with write access specified. I'm also going to modify the return value to return the CanWrite flag too just for extra sanity checking.

Reply:
Wouldn't that be an issue for dirs without a cs file?